core/src/test/java/io/questdb/test/griffin/engine/groupby/hyperloglog/HyperLogLogTest.java

```diff
@@ -0,0 +1,361 @@
+import io.questdb.std.Hash;
+import io.questdb.std.IntHashSet;
+import io.questdb.std.Rnd;
+import io.questdb.test.AbstractTest;
+import io.questdb.test.tools.TestUtils;
+import org.junit.Test;
+
+import static io.questdb.test.griffin.engine.groupby.hyperloglog.HyperLogLogTestUtils.assertCardinality;
+import static io.questdb.test.tools.TestUtils.assertMemoryLeak;
+import static org.junit.Assert.*;
+
+public class HyperLogLogTest extends AbstractTest {
+
+    @Test
+    public void testAddInvalidatesCachedCardinality() throws Exception {
+        CairoConfiguration config = new DefaultCairoConfiguration(root) {
+            @Override
+            public long getGroupByAllocatorDefaultChunkSize() {
+                return 64;
+            }
+        };
+        for (int precision = 4; precision <= 18; precision++) {
+            int finalPrecision = precision;
+            assertMemoryLeak(() -> {
+                try (GroupByAllocator allocator = new GroupByAllocator(config)) {
+                    Rnd rnd = new Rnd();
+
+                    HyperLogLog hll = new HyperLogLog(finalPrecision);
+                    hll.setAllocator(allocator);
+                    hll.of(0);
+
+                    hll.addAndComputeCardinalityFast(Hash.murmur3ToLong(rnd.nextLong()));
+                    assertEquals(1, hll.computeCardinality());
+                    assertEquals(1, hll.computeCardinality());
+
+                    // add should invalidate cache
+                    hll.addAndComputeCardinalityFast(Hash.murmur3ToLong(rnd.nextLong()));
+                    assertEquals(2, hll.computeCardinality());
+                }
+            });
+        }
+    }
+
+    @Test
+    public void testFastCardinalityComputation() throws Exception {
+        CairoConfiguration config = new DefaultCairoConfiguration(root) {
+            @Override
+            public long getGroupByAllocatorDefaultChunkSize() {
+                return 64;
+            }
+        };
+        for (int precision = 4; precision <= 18; precision++) {
+            int finalPrecision = precision;
+            assertMemoryLeak(() -> {
+                try (GroupByAllocator allocator = new GroupByAllocator(config)) {
+                    HyperLogLog hll = new HyperLogLog(finalPrecision);
+                    hll.setAllocator(allocator);
+                    hll.of(0);
+
+                    int exactCardinality = 10000;
+
+                    for (int i = 0; i < exactCardinality; i++) {
+                        long estimated = hll.addAndComputeCardinalityFast(Hash.murmur3ToLong(i));
+                        if (hll.isSparse()) {
+                            assertCardinality(i + 1, finalPrecision, estimated);
+                        } else {
+                            assertEquals(-1, estimated);
+                        }
+                    }
+                }
+            });
+        }
+    }
+
+    @Test
+    public void testFuzz() throws Exception {
```

Comment:
This test takes 13+ secs on my machine. On CI, it'll be even slower. We need to speed it up. An option to do that might be to use random precision from the allowed range instead of looping through all of them.

Reply:
Replacing the loop with randomly selecting precision reduced the execution time from 15 seconds to 1 second on my desktop.

Reply:
Thanks for the speed up.